### PR TITLE
feat(cli): auto setup typescript during start

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Some users are confused about how to use the fully automated TypeScript setup. It currently runs when you run `npx expo start` in a project with `tsconfig.json` or any TypeScript file in the project.
- This change will run the TypeScript check continuously until TypeScript is setup. This only applies to projects using Metro, and currently is only run in `npx expo start`, and the run commands (when the server is started in the current window, e.g. not headless mode).

# How

- Fork the `runServer` function from Metro into the CLI so we can access the Metro bundler instance directly.
- Use Metro's native file watching to observe new TypeScript files being added to the project (within the observable files). This is fast and ensures we don't accidentally block a thread that otherwise could've been used for bundling.
- The TypeScript continuous check only runs once, if you remove TypeScript and add it back in the same process, then nothing happens. This is a performance optimization.
- If TypeScript is enabled while the process is running, the user will not be prompted to install, the packages will just install in the project. Failures will be displayed in the running process, and no more attempts will be made.
- If TypeScript is detected on start then the existing system will be used and the continuous check will be skipped, this means projects already using TypeScript won't have any technical performance hit.

# Test Plan

- In a project without TypeScript:
  - `npx expo`
  - `touch tsconfig.json` or `mv index.js index.tsx` or `touch foo.tsx`
  - TypeScript is automatically installed for the project.
  - `EXPO_NO_TYPESCRIPT_SETUP` voids this behavior
- Added unit tests.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts [hunk 1]
```diff
@@ -5,20 +5,29 @@ import path from 'path';
...
+import { memoize } from '../../../utils/fn';
 import { everyMatchAsync, wrapGlobWithTimeout } from '../../../utils/glob';
 import { ProjectPrerequisite } from '../Prerequisite';
 import { ensureDependenciesAsync } from '../dependencies/ensureDependenciesAsync';
 import { updateTSConfigAsync } from './updateTSConfig';
 
 const debug = require('debug')('expo:doctor:typescriptSupport') as typeof console.log;
 
+const warnDisabled = memoize(() => {
+  Log.warn('Skipping TypeScript setup: EXPO_NO_TYPESCRIPT_SETUP is enabled.');
```
Comment (byCedric) [inline packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts:17]: Is this being wrapped inside memoize only to warn once?

File: packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts [hunk 2]
```diff
@@ -5,20 +5,29 @@ import path from 'path';
...
+});
+
 /** Ensure the project has the required TypeScript support settings. */
-export class TypeScriptProjectPrerequisite extends ProjectPrerequisite {
-  /** Ensure a project that hasn't explicitly disabled web support has all the required packages for running in the browser. */
-  async assertImplementation(): Promise<void> {
+export class TypeScriptProjectPrerequisite extends ProjectPrerequisite<boolean> {
+  /**
+   * Ensure a project that hasn't explicitly disabled web support has all the required packages for running in the browser.
```
Comment (byCedric) [inline packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts:23]:
```suggestion
   * Ensure a project that hasn't explicitly disabled typescript support has all the required packages for running in the browser.
```

**Review by byCedric — APPROVED** [review]
This looks like a fundamentally better approach compared to jest workers. Just a few TS issues to resolve.

Also, just for administration, I assume this supersedes:
  - https://github.com/expo/expo/pull/20883
  - https://github.com/expo/expo/pull/21466

File: packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts (hunk 1 above)
Comment (EvanBacon) [reply packages/@expo/cli/src/start/doctor/typescript/TypeScriptProjectPrerequisite.ts:17]: yes

**Pull request merged**